src/node_zip.cc

```diff
@@ -0,0 +1,454 @@
+  CHECK_LE(file_stat.comp_size, ZIP_INT64_MAX);
+  CHECK_LE(file_stat.comp_size, std::numeric_limits<size_t>::max());
+
+  char* data = reinterpret_cast<char*>(malloc(file_stat.comp_size * sizeof(char)));
+
+  zip_file_t* file;
+  CHECK_NE(file = zip_fopen_index(zip->zip_, index, ZIP_FL_COMPRESSED), nullptr);
+
+  zip_int64_t read_length = zip_fread(file, data, ZIP_INT64_MAX);
+  CHECK_EQ(read_length, static_cast<zip_int64_t>(file_stat.comp_size));
+
+  v8::Local<v8::Value> result[] = {
+    v8::Integer::NewFromUnsigned(isolate, file_stat.comp_method),
+    Buffer::New(isolate, data, read_length).ToLocalChecked(),
+  };
+
+  args.GetReturnValue().Set(v8::Array::New(isolate, &result[0], sizeof(result) / sizeof(result[0])));
+}
+
+void ZipArchive::StatEntry(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+
+  ZipArchive* zip;
+  zip_uint32_t index;
+  ZipArchive::CheckZipEntry(args, zip, index);
+
+  zip_stat_t file_stat;
+  CHECK_EQ(zip_stat_index(zip->zip_, index, 0, &file_stat), 0);
+  CHECK_LE(file_stat.comp_size, ZIP_INT64_MAX);
+  CHECK_LE(file_stat.comp_size, std::numeric_limits<size_t>::max());
+
+  zip_uint8_t opsys;
+  zip_uint32_t attributes;
+  CHECK_EQ(zip_file_get_external_attributes(zip->zip_, index, 0, &opsys, &attributes), 0);
+
+  v8::Local<v8::Value> result[] = {
+    v8::Integer::NewFromUnsigned(isolate, file_stat.size),
+    v8::Integer::NewFromUnsigned(isolate, file_stat.comp_size),
+    v8::Integer::NewFromUnsigned(isolate, file_stat.comp_method),
+    v8::Integer::NewFromUnsigned(isolate, file_stat.mtime),
+    v8::Integer::NewFromUnsigned(isolate, file_stat.crc),
+    v8::Integer::NewFromUnsigned(isolate, opsys),
+    v8::Integer::NewFromUnsigned(isolate, attributes),
+  };
+
+  args.GetReturnValue().Set(v8::Array::New(isolate, &result[0], sizeof(result) / sizeof(result[0])));
+}
+
+void ZipArchive::RestatEntry(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+  Environment* env = Environment::GetCurrent(isolate);
+  v8::Local<v8::Context> ctx = env->context();
+
+  ZipArchive* zip;
+  zip_uint32_t index;
+  ZipArchive::CheckZipEntry(args, zip, index);
+
+  const int argc = args.Length();
+  CHECK_GE(argc, 2);
+
+  CHECK(args[1]->IsObject());
+  v8::Local<v8::Object> stats = args[1]->ToObject(ctx).ToLocalChecked();
+
+  v8::Local<v8::Value> mtime;
+  if (stats->Get(ctx, v8::String::NewFromUtf8Literal(isolate, "mtime")).ToLocal(&mtime)) {
+    time_t mtime_val = mtime->ToInteger(ctx).ToLocalChecked()->Value();
+
+    CHECK_EQ(zip_file_set_mtime(zip->zip_, index, mtime_val, 0), 0);
+  }
+
+  v8::Local<v8::Value> opsys;
+  if (stats->Get(ctx, v8::String::NewFromUtf8Literal(isolate, "opsys")).ToLocal(&opsys)) {
+    auto attributes = stats->Get(ctx, v8::String::NewFromUtf8Literal(isolate, "attributes"))
+      .FromMaybe<v8::Value>(v8::Integer::New(isolate, 0))->ToInteger(ctx).ToLocalChecked();
```

Comment:
Similarly here, you would not want to use `.ToLocalChecked()` and instead just `return;` if `.ToLocal()` returns `false`, since converting a JS property value to an integer can throw exceptions